Malformed request body on application start

- Caused by #60, #62 

### Steps to reproduce

- `$ sbt run`

### Observed

An exception

```
org.http4s.MalformedMessageBodyFailure: Malformed request body: Invalid JSON
	at org.http4s.jawn.JawnInstances$$anonfun$jawnDecoder$1$$anonfun$apply$1.applyOrElse(JawnInstances.scala:16) ~[http4s-jawn_2.11-0.15.6.jar:0.15.6]
	at org.http4s.jawn.JawnInstances$$anonfun$jawnDecoder$1$$anonfun$apply$1.applyOrElse(JawnInstances.scala:14) ~[http4s-jawn_2.11-0.15.6.jar:0.15.6]
	at scala.PartialFunction$Lifted.apply(PartialFunction.scala:223) ~[scala-library-2.11.8.jar:1.0.0-M1]
	at scala.PartialFunction$Lifted.apply(PartialFunction.scala:219) ~[scala-library-2.11.8.jar:1.0.0-M1]
	at scalaz.stream.Process$$anonfun$partialAttempt$1.apply(Process.scala:392) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$partialAttempt$1.apply(Process.scala:392) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$attempt$1$$anonfun$apply$22.apply(Process.scala:230) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$attempt$1$$anonfun$apply$22.apply(Process.scala:230) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Util$.Try(Util.scala:38) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$attempt$1.apply(Process.scala:230) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$attempt$1.apply(Process.scala:229) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$2$$anonfun$apply$12$$anonfun$apply$13.apply(Process.scala:115) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$2$$anonfun$apply$12$$anonfun$apply$13.apply(Process.scala:115) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Util$.Try(Util.scala:38) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$2$$anonfun$apply$12.apply(Process.scala:115) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$2$$anonfun$apply$12.apply(Process.scala:115) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.Trampoline$$anonfun$delay$1.apply(Free.scala:262) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Trampoline$$anonfun$delay$1.apply(Free.scala:262) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$run$1.apply(Free.scala:187) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$run$1.apply(Free.scala:187) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free.go2$1(Free.scala:134) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free.go(Free.scala:137) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free.run(Free.scala:187) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:84) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:84) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Util$.Try(Util.scala:38) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$class.go$1(Process.scala:84) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$class.step(Process.scala:97) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$Append.step(Process.scala:641) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$pipe$1.apply(Process.scala:143) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$pipe$1.apply(Process.scala:139) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$flatMap$1.apply(Process.scala:46) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$flatMap$1.apply(Process.scala:46) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Util$.Try(Util.scala:38) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$class.flatMap(Process.scala:46) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$Emit.flatMap(Process.scala:592) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$flatMap$4.apply(Process.scala:48) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$flatMap$4.apply(Process.scala:48) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.Free$$anonfun$map$1.apply(Free.scala:53) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$map$1.apply(Free.scala:53) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free.scalaz$Free$$fastFlatMap(Free.scala:72) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$scalaz$Free$$fastFlatMap$1.apply(Free.scala:73) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$scalaz$Free$$fastFlatMap$1.apply(Free.scala:73) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.std.FunctionInstances$$anon$1$$anonfun$map$1.apply(Function.scala:56) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$run$1.apply(Free.scala:187) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$run$1.apply(Free.scala:187) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free.go2$1(Free.scala:134) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free.go(Free.scala:137) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free.run(Free.scala:187) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:84) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:84) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Util$.Try(Util.scala:38) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$class.go$1(Process.scala:84) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$class.step(Process.scala:97) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$Append.step(Process.scala:641) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$class.go$3(Process.scala:482) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$go$3$3.apply(Process.scala:490) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$go$3$3.apply(Process.scala:490) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.concurrent.Task$$anonfun$flatMap$1$$anonfun$1.apply(Task.scala:36) ~[scalaz-concurrent_2.11-7.1.11.jar:na]
	at scalaz.concurrent.Task$$anonfun$flatMap$1$$anonfun$1.apply(Task.scala:36) ~[scalaz-concurrent_2.11-7.1.11.jar:na]
	at scalaz.concurrent.Task$.Try(Task.scala:403) ~[scalaz-concurrent_2.11-7.1.11.jar:na]
	at scalaz.concurrent.Task$$anonfun$flatMap$1.apply(Task.scala:36) ~[scalaz-concurrent_2.11-7.1.11.jar:na]
	at scalaz.concurrent.Task$$anonfun$flatMap$1.apply(Task.scala:34) ~[scalaz-concurrent_2.11-7.1.11.jar:na]
	at scala.Function1$$anonfun$andThen$1.apply(Function1.scala:52) ~[scala-library-2.11.8.jar:1.0.0-M1]
	at scala.Function1$$anonfun$andThen$1.apply(Function1.scala:52) ~[scala-library-2.11.8.jar:1.0.0-M1]
	at scalaz.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:58) ~[scalaz-concurrent_2.11-7.1.11.jar:na]
	at scalaz.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:58) ~[scalaz-concurrent_2.11-7.1.11.jar:na]
	at scalaz.concurrent.Future.step(Future.scala:109) ~[scalaz-concurrent_2.11-7.1.11.jar:na]
	at scalaz.concurrent.Future.listen(Future.scala:75) ~[scalaz-concurrent_2.11-7.1.11.jar:na]
	at scalaz.concurrent.Future$$anonfun$listen$1$$anonfun$apply$4.apply(Future.scala:79) ~[scalaz-concurrent_2.11-7.1.11.jar:na]
	at scalaz.concurrent.Future$$anonfun$listen$1$$anonfun$apply$4.apply(Future.scala:79) ~[scalaz-concurrent_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$map$1.apply(Free.scala:53) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$map$1.apply(Free.scala:53) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free.scalaz$Free$$fastFlatMap(Free.scala:72) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$resume$1.apply(Free.scala:88) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$resume$1.apply(Free.scala:88) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.std.FunctionInstances$$anon$1$$anonfun$map$1.apply(Function.scala:56) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$run$1.apply(Free.scala:187) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$run$1.apply(Free.scala:187) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free.go2$1(Free.scala:134) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free.go(Free.scala:137) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free.run(Free.scala:187) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.concurrent.Future$$anonfun$apply$15$$anon$4.call(Future.scala:380) [scalaz-concurrent_2.11-7.1.11.jar:na]
	at scalaz.concurrent.Future$$anonfun$apply$15$$anon$4.call(Future.scala:380) [scalaz-concurrent_2.11-7.1.11.jar:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_76]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_76]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_76]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_76]
Caused by: jawn.ParseException: exhausted input
	at jawn.AsyncParser.churn(AsyncParser.scala:234) ~[jawn-parser_2.11-0.10.4.jar:0.10.4]
	at jawn.AsyncParser.finish(AsyncParser.scala:98) ~[jawn-parser_2.11-0.10.4.jar:0.10.4]
	at jawnstreamz.package$$anonfun$parseJson$1$$anonfun$apply$2.apply(package.scala:31) ~[jawn-streamz_2.11-0.10.1.jar:0.10.1]
	at jawnstreamz.package$$anonfun$parseJson$1$$anonfun$apply$2.apply(package.scala:31) ~[jawn-streamz_2.11-0.10.1.jar:0.10.1]
	at scalaz.stream.Process$$anonfun$onComplete$1.apply(Process.scala:338) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$onComplete$1.apply(Process.scala:338) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$2$$anonfun$apply$12$$anonfun$apply$13.apply(Process.scala:115) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$2$$anonfun$apply$12$$anonfun$apply$13.apply(Process.scala:115) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Util$.Try(Util.scala:38) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$2$$anonfun$apply$12.apply(Process.scala:115) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$2$$anonfun$apply$12.apply(Process.scala:115) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.Trampoline$$anonfun$delay$1.apply(Free.scala:262) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Trampoline$$anonfun$delay$1.apply(Free.scala:262) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$run$1.apply(Free.scala:187) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free$$anonfun$run$1.apply(Free.scala:187) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free.go2$1(Free.scala:134) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free.go(Free.scala:137) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Free.run(Free.scala:187) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:84) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:84) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Util$.Try(Util.scala:38) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$class.go$1(Process.scala:84) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$class.step(Process.scala:97) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$Append.step(Process.scala:641) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$class.disconnect(Process.scala:247) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$Append.disconnect(Process.scala:641) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$disconnect$2.apply(Process.scala:249) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$disconnect$2.apply(Process.scala:249) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$suspend$1.apply(Process.scala:1498) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$suspend$1.apply(Process.scala:1497) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:84) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:84) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Util$.Try(Util.scala:38) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$class.go$1(Process.scala:84) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$class.step(Process.scala:97) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$Append.step(Process.scala:641) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$suspendStep$1.apply(Process.scala:106) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$suspendStep$1.apply(Process.scala:105) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$2$$anonfun$apply$12$$anonfun$apply$13.apply(Process.scala:115) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$2$$anonfun$apply$12$$anonfun$apply$13.apply(Process.scala:115) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Util$.Try(Util.scala:38) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$2$$anonfun$apply$12.apply(Process.scala:115) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.stream.Process$$anonfun$2$$anonfun$apply$12.apply(Process.scala:115) ~[scalaz-stream_2.11-0.8.6.jar:na]
	at scalaz.Trampoline$$anonfun$delay$1.apply(Free.scala:262) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.Trampoline$$anonfun$delay$1.apply(Free.scala:262) ~[scalaz-core_2.11-7.1.11.jar:na]
	at scalaz.std.FunctionInstances$$anon$1$$anonfun$map$1.apply(Function.scala:56) ~[scalaz-core_2.11-7.1.11.jar:na]
	... 70 common frames omitted
```

### Expected

No exceptions

--- a/src/main/scala/me/rexim/issuestant/github/EtagPolling.scala
+++ b/src/main/scala/me/rexim/issuestant/github/EtagPolling.scala
@@ -15,7 +15,7 @@ import scalaz.stream._
 import scalaz.concurrent._
 
 class EtagPolling[E](client: Client, pollingUri: Uri) {
-  // TODO: make EtagPolling[E].Item a class, not a tuple
+  // TODO(#65): make EtagPolling[E].Item a class, not a tuple
   type Item = (Option[String], Option[E])
 
   def responses(implicit decoder: Decoder[E]): Process[Task, E] =

--- a/src/main/scala/me/rexim/issuestant/github/EtagPolling.scala
+++ b/src/main/scala/me/rexim/issuestant/github/EtagPolling.scala
@@ -15,10 +15,13 @@ import scalaz.stream._
 import scalaz.concurrent._
 
 class EtagPolling[E](client: Client, pollingUri: Uri) {
+  // TODO: make EtagPolling[E].Item a class, not a tuple
   type Item = (Option[String], Option[E])
 
-  def responses(implicit decoder: Decoder[E]): Process[Task, Item] =
-    Process.iterateEval((None, None): Item)(pollingIteration(_)(decoder)).tail
+  def responses(implicit decoder: Decoder[E]): Process[Task, E] =
+    Process
+      .iterateEval((None, None): Item)(pollingIteration(_)(decoder))
+      .collect { case (_, Some(e)) => e }
 
   private def getETag(response: Response): Option[String] =
     response.headers.get(CaseInsensitiveString("ETag")).map(_.value)

--- a/src/main/scala/me/rexim/issuestant/github/EtagPolling.scala
+++ b/src/main/scala/me/rexim/issuestant/github/EtagPolling.scala
@@ -1,29 +1,47 @@
 package me.rexim.issuestant.github
 
+import io.circe._
+import io.circe.generic.auto._
+import io.circe.parser._
+import io.circe.syntax._
+
 import org.http4s._
 import org.http4s.client._
+import org.http4s.circe._
 import org.http4s.util._
 
 import scalaz._
 import scalaz.stream._
 import scalaz.concurrent._
 
-class EtagPolling(client: Client, pollingUri: Uri) {
-  def responses: Process[Task, Response] =
-    Process.iterateEval(Response())(pollingIteration).tail.filter(_.status == Status.Ok)
+class EtagPolling[E](client: Client, pollingUri: Uri) {
+  type Item = (Option[String], Option[E])
+
+  def responses(implicit decoder: Decoder[E]): Process[Task, Item] =
+    Process.iterateEval((None, None): Item)(pollingIteration(_)(decoder)).tail
 
   private def getETag(response: Response): Option[String] =
     response.headers.get(CaseInsensitiveString("ETag")).map(_.value)
 
-  private def nextRequest(previousResponse: Response): Request =
+  private def nextRequest(previousItem: Item): Request =
     Request (
       uri = pollingUri,
       headers =
-        getETag(previousResponse)
+        previousItem._1
           .map(e => Headers(Header("If-None-Match", e)))
           .getOrElse(Headers.empty)
     )
 
-  private def pollingIteration(previousResponse: Response): Task[Response] =
-    client.fetch[Response](nextRequest(previousResponse))((x) => Task(x))
+  private def pollingIteration(previousItem: Item)(implicit decoder: Decoder[E]): Task[Item] =
+    client.fetch[Item](nextRequest(previousItem)) { (response) =>
+      implicit val http4sDecoder = jsonOf[E]
+
+      if (response.status == Status.Ok) {
+        response.as[E].map { e =>
+          (getETag(response), Some(e))
+        }
+      } else {
+        Task((getETag(response), None))
+      }
+    }
 }

--- a/src/main/scala/me/rexim/issuestant/github/EventsSource.scala
+++ b/src/main/scala/me/rexim/issuestant/github/EventsSource.scala
@@ -33,7 +33,7 @@ import grizzled.slf4j.Logging
   */
 // $COVERAGE-OFF$
 class EventsSource[E] (etagPolling: EtagPolling[List[E]]) extends Logging {
-  // TODO: EventsSource should filter out events that already happened
+  // TODO(#66): EventsSource should filter out events that already happened
   def events(implicit decoder: Decoder[E]): Process[Task, E] =
     etagPolling.responses
       .flatMap(Process.emitAll)

--- a/src/main/scala/me/rexim/issuestant/github/EventsSource.scala
+++ b/src/main/scala/me/rexim/issuestant/github/EventsSource.scala
@@ -32,19 +32,18 @@ import grizzled.slf4j.Logging
   * @param repo the name of the repo
   */
 // $COVERAGE-OFF$
-class EventsSource[E] (etagPolling: EtagPolling) extends Logging {
+class EventsSource[E] (etagPolling: EtagPolling[List[E]]) extends Logging {
 
   def events(implicit decoder: Decoder[E]): Process[Task, E] =
     etagPolling.responses
       .flatMap(extractEvents)
       .map((e) => { info(s"New GitHub event: ${e}"); e })
 
-  private def extractEvents(response: Response)
-    (implicit decoder: Decoder[E]): Process[Task, E] = {
-    implicit val eventHttp4s = jsonOf[E]
-    implicit val listEventHttp4s = jsonOf[List[E]]
-
-    Process.emitAll(response.as[List[E]].run)
+  private def extractEvents(item: (Option[String], Option[List[E]])): Process[Task, E] = {
+    item match {
+      case (_, Some(events)) => Process.emitAll(events)
+      case _ => Process.empty
+    }
   }
 }
 // $COVERAGE-ON$

--- a/src/main/scala/me/rexim/issuestant/github/EventsSource.scala
+++ b/src/main/scala/me/rexim/issuestant/github/EventsSource.scala
@@ -33,17 +33,10 @@ import grizzled.slf4j.Logging
   */
 // $COVERAGE-OFF$
 class EventsSource[E] (etagPolling: EtagPolling[List[E]]) extends Logging {
-
+  // TODO: EventsSource should filter out events that already happened
   def events(implicit decoder: Decoder[E]): Process[Task, E] =
     etagPolling.responses
-      .flatMap(extractEvents)
+      .flatMap(Process.emitAll)
       .map((e) => { info(s"New GitHub event: ${e}"); e })
-
-  private def extractEvents(item: (Option[String], Option[List[E]])): Process[Task, E] = {
-    item match {
-      case (_, Some(events)) => Process.emitAll(events)
-      case _ => Process.empty
-    }
-  }
 }
 // $COVERAGE-ON$

--- a/src/test/scala/me/rexim/github/EtagPollingSpec.scala
+++ b/src/test/scala/me/rexim/github/EtagPollingSpec.scala
@@ -34,7 +34,7 @@ class EtagPollingSpec extends FlatSpec with Matchers {
 
     requests.map(_.uri) should be ((1 to requestCount).map(_ => requestUri))
 
-    (None :: responses.map(_._1)) should
+    (None :: responses.map(Some(_)).init) should
       be (requests.map(_.headers.get(ifNoneMatchName).map(_.value)))
   }
 }

--- a/src/test/scala/me/rexim/github/EtagPollingSpec.scala
+++ b/src/test/scala/me/rexim/github/EtagPollingSpec.scala
@@ -1,5 +1,10 @@
 package me.rexim.issuestant.github
 
+import io.circe._
+import io.circe.generic.auto._
+import io.circe.parser._
+import io.circe.syntax._
+
 import org.scalatest._
 import me.rexim.issuestant.mock._
 
@@ -19,7 +24,7 @@ class EtagPollingSpec extends FlatSpec with Matchers {
     val requestUri = Uri(path = "/rexim")
 
     val etagLogging = new EtagLogging
-    val etagPolling = new EtagPolling(
+    val etagPolling = new EtagPolling[String](
       etagLogging.client,
       requestUri
     )
@@ -29,7 +34,7 @@ class EtagPollingSpec extends FlatSpec with Matchers {
 
     requests.map(_.uri) should be ((1 to requestCount).map(_ => requestUri))
 
-    (None :: responses.map(_.headers.get(etagName).map(_.value)).init) should
+    (None :: responses.map(_._1)) should
       be (requests.map(_.headers.get(ifNoneMatchName).map(_.value)))
   }
 }

--- a/src/test/scala/me/rexim/mock/EtagLogging.scala
+++ b/src/test/scala/me/rexim/mock/EtagLogging.scala
@@ -7,6 +7,9 @@ import org.http4s._
 import org.http4s.client._
 
 import scalaz.concurrent._
+import scalaz.stream._
+
+import scodec.bits._
 
 // TODO(#55): Make ETagLogging less stateful
 //
@@ -31,7 +34,8 @@ class EtagLogging {
           status = Status.Ok,
           headers = Headers(
             Header("ETag", randomEtag())
-          )
+          ),
+          body = Process.emit(ByteVector("\"hello\"".getBytes))
         )
       }
     }.map((r) => DisposableResponse(r, Task({}))),

--- a/src/test/scala/me/rexim/mock/EtagLogging.scala
+++ b/src/test/scala/me/rexim/mock/EtagLogging.scala
@@ -29,13 +29,14 @@ class EtagLogging {
   lazy val client: Client = Client(
     open = HttpService.lift { (request) =>
       Task {
+        val etag = randomEtag()
         requestLog += request
         Response(
           status = Status.Ok,
           headers = Headers(
-            Header("ETag", randomEtag())
+            Header("ETag", etag)
           ),
-          body = Process.emit(ByteVector("\"hello\"".getBytes))
+          body = Process.emit(ByteVector(s"""\"$etag\"""".getBytes))
         )
       }
     }.map((r) => DisposableResponse(r, Task({}))),

